Support ppxlib.0.36.0

patricoferris opened the pull request:
This is a draft PR bump to `bisect_ppx` to the latest ppxlib.

I think this is realistically going to need #445 and I'll need to look into some of the failing tests (some are formatting/pretty-printing related I think).

patricoferris committed:
Support ppxlib.0.36.0

glondu commented:
Any news?

Kakadu commented:
@aantron ^^

NathanReb commented:
I think this patch is causing #450. There's something fishy going on when instrumenting `Pfunction_cases` that can lead to it being converted to a `Pfunction_body` which triggers the compiler invariant in the linked issue.

I'll look into it and push a fix!

Once that's done I'm also happy to help getting a release in any way I can @aantron !

NathanReb committed:
Add regression test for function_cases instrumentation bug

Signed-off-by: Nathan Rebours <nathan.rebours@ocamlpro.com>

NathanReb committed:
Fix instrumentation of [function _ ->] expressions

Signed-off-by: Nathan Rebours <nathan.rebours@ocamlpro.com>

NathanReb commented:
Opened a PR to @patricoferris' branch here: https://github.com/patricoferris/bisect_ppx/pull/1

NathanReb commented:
The test suite is broken at the moment so I tried to make the minimal set of changes required to make my new regression test work, for anybody willing to try it out, you can run just the one test with:
```
dune runtest test/instrument/function.t
```
to get a clean output that's not polluted by the other errors.

I'd also be happy to contribute a refresh of the test suite and to help with the project maintenance. Don't hesitate to reach out if you're looking for a bit of help on that front!

patricoferris commented:
Pulled your changes in @NathanReb -- thanks !

aantron commented:
I took a look at all the background today and I'm going to continue with reviewing the code in the coming days.

aantron commented:
Links for the future:

- The OCaml RFC that caused this https://github.com/ocaml/RFCs/pull/32
- The OCaml PR https://github.com/ocaml/ocaml/pull/12236, in particular https://github.com/ocaml/ocaml/commit/1fdfc412e4f8735cf27d5fa176fedbea49d8b132 and https://github.com/ocaml/ocaml/commit/437548629a74408515ad0059d417fe2091d9715b
- Ppxlib PR https://github.com/ocaml-ppx/ppxlib/pull/451

kit-ty-kate commented:
FYI: OCaml 5.4.0 was released yesterday

aantron committed:
opam: update ppxlib constraint

aantron committed:
Merge branch 'master' into 5.2

aantron commented:
The test suite presently works for `master` on 4.08-4.13 with ppxlib 0.35.0, including [in CI](https://github.com/aantron/bisect_ppx/actions/runs/18446320437/job/52553541766). I just merged `master` into this PR, to have the CI test ppxlib's translation of what this PR does to older versions. Could you pull these commits down and comment on the [failures](https://github.com/aantron/bisect_ppx/actions/runs/18446347181/job/52553605885?pr=448)?

```diff
dune build -p bisect_ppx
NO_COLOR=yes dune build -p bisect_ppx @runtest
File "test/instrument/control/function.t", line 1, characters 0-0:
/usr/bin/git --no-pager diff --no-index --color=always -u _build/default/test/instrument/control/function.t _build/default/test/instrument/control/function.t.corrected
diff --git a/_build/default/test/instrument/control/function.t b/_build/default/test/instrument/control/function.t.corrected
index bfb978d..a39693a 100644
--- a/_build/default/test/instrument/control/function.t
+++ b/_build/default/test/instrument/control/function.t.corrected
@@ -59,6 +59,7 @@ Or-pattern.
   > EOF
   let _ =
    fun ___bisect_matched_value___ ->
+    ___bisect_visit___ 2;
     match ___bisect_matched_value___ with
     | None | Some _ ->
         (match[@ocaml.warning "-4-8-9-11-26-27-28-33"]
File "test/instrument/control/newtype.t", line 1, characters 0-0:
/usr/bin/git --no-pager diff --no-index --color=always -u _build/default/test/instrument/control/newtype.t _build/default/test/instrument/control/newtype.t.corrected
diff --git a/_build/default/test/instrument/control/newtype.t b/_build/default/test/instrument/control/newtype.t.corrected
index 630a122..12927d2 100644
--- a/_build/default/test/instrument/control/newtype.t
+++ b/_build/default/test/instrument/control/newtype.t.corrected
@@ -28,6 +28,6 @@ Subexpression in tail position iff whole expression is in tail position.
   let _ = fun (type _t) -> ___bisect_post_visit___ 0 (print_endline "foo")
   
   let _ =
-   fun () ->
+   fun () (type _t) ->
     ___bisect_visit___ 1;
-    fun (type _t) -> print_endline "foo"
+    print_endline "foo"
File "test/instrument/pattern/exception.t", line 1, characters 0-0:
/usr/bin/git --no-pager diff --no-index --color=always -u _build/default/test/instrument/pattern/exception.t _build/default/test/instrument/pattern/exception.t.corrected
diff --git a/_build/default/test/instrument/pattern/exception.t b/_build/default/test/instrument/pattern/exception.t.corrected
index 5c660dc..e383838 100644
--- a/_build/default/test/instrument/pattern/exception.t
+++ b/_build/default/test/instrument/pattern/exception.t.corrected
@@ -38,10 +38,12 @@ patterns.
   >   | _ -> print_endline "baz"
   > EOF
   let _ =
-    let ___bisect_case_0___ x () =
+    let ___bisect_case_0___ x =
+     fun [@ppxlib.migration.stop_taking] () ->
       ignore x;
       ___bisect_post_visit___ 0 (print_endline "foo")
-    and ___bisect_case_1___ y () =
+    and ___bisect_case_1___ y =
+     fun [@ppxlib.migration.stop_taking] () ->
       ignore y;
       ___bisect_post_visit___ 1 (print_endline "bar")
     in
File "test/instrument/recent/gadt.t", line 1, characters 0-0:
/usr/bin/git --no-pager diff --no-index --color=always -u _build/default/test/instrument/recent/gadt.t _build/default/test/instrument/recent/gadt.t.corrected
diff --git a/_build/default/test/instrument/recent/gadt.t b/_build/default/test/instrument/recent/gadt.t.corrected
index d65beb4..789f6d4 100644
--- a/_build/default/test/instrument/recent/gadt.t
+++ b/_build/default/test/instrument/recent/gadt.t.corrected
@@ -39,6 +39,7 @@ With function.
   
   let f : type a. a t -> unit =
    fun ___bisect_matched_value___ ->
+    ___bisect_visit___ 2;
     match ___bisect_matched_value___ with
     | A | B ->
         (match[@ocaml.warning "-4-8-9-11-26-27-28-33"]
File "test/instrument/value.t", line 1, characters 0-0:
/usr/bin/git --no-pager diff --no-index --color=always -u _build/default/test/instrument/value.t _build/default/test/instrument/value.t.corrected
diff --git a/_build/default/test/instrument/value.t b/_build/default/test/instrument/value.t.corrected
index 2d9e0fa..58562b7 100644
--- a/_build/default/test/instrument/value.t
+++ b/_build/default/test/instrument/value.t.corrected
@@ -194,7 +194,11 @@ No instrumentation is inserted into expressions that are (syntactic) values.
   
   let _ = (___bisect_post_visit___ 1 (f ()) :> [ `Foo | `Bar ])
   
-  let _ = fun () -> (f () :> [ `Foo | `Bar ])
+  let _ =
+   fun () ->
+    (___bisect_visit___ 2;
+     f ()
+      :> [ `Foo | `Bar ])
 
 
   $ bash test.sh <<'EOF'
make: *** [Makefile:13: test] Error 1
```

aantron commented:
This difference

```diff
diff --git a/_build/default/test/instrument/value.t b/_build/default/test/instrument/value.t.corrected
index 2d9e0fa..08f2932 100644
--- a/_build/default/test/instrument/value.t
+++ b/_build/default/test/instrument/value.t.corrected
@@ -194,7 +194,10 @@ No instrumentation is inserted into expressions that are (syntactic) values.
   
   let _ = (___bisect_post_visit___ 1 (f ()) :> [ `Foo | `Bar ])
   
-  let _ = fun () -> (f () :> [ `Foo | `Bar ])
+  let _ =
+   fun () ->
+    ___bisect_visit___ 2;
+    (f () :> [ `Foo | `Bar ])
 
 
   $ bash test.sh <<'EOF'
```

appears in the 4.14 and 5.0.0 run in `master`, without this PR, so probably not caused by it. IIRC I have a commit in a branch locally that suppresses this, so I'll push that in later. Don't spend time addressing this one :)

aantron commented:
> I'd also be happy to contribute a refresh of the test suite and to help with the project maintenance. Don't hesitate to reach out if you're looking for a bit of help on that front!

Would be very glad to merge updates to the test suite, and help with project maintenance. What would you like to do? Do you need write access to the repo?

patricoferris commented:
Thanks @aantron.

```diff
-    let ___bisect_case_0___ x () =
+    let ___bisect_case_0___ x =
+     fun [@ppxlib.migration.stop_taking] () ->
       ignore x;
       ___bisect_post_visit___ 0 (print_endline "foo")
-    and ___bisect_case_1___ y () =
+    and ___bisect_case_1___ y =
+     fun [@ppxlib.migration.stop_taking] () ->
       ignore y;
       ___bisect_post_visit___ 1 (print_endline "bar")
     in
 ```
 
 these failures are a known bug that is fixed upstream, see https://github.com/ocaml-ppx/ppxlib/pull/604.
 
 The others look like some interaction with the new function layout in the parsetree (which I may have not implemented correctly).
 
 ```diff
 diff --git a/_build/default/test/instrument/control/newtype.t b/_build/default/test/instrument/control/newtype.t.corrected
index 630a122..12927d2 100644
--- a/_build/default/test/instrument/control/newtype.t
+++ b/_build/default/test/instrument/control/newtype.t.corrected
@@ -28,6 +28,6 @@ Subexpression in tail position iff whole expression is in tail position.
   let _ = fun (type _t) -> ___bisect_post_visit___ 0 (print_endline "foo")
   
   let _ =
-   fun () ->
+   fun () (type _t) ->
     ___bisect_visit___ 1;
-    fun (type _t) -> print_endline "foo"
+    print_endline "foo"
```

This looks like we go all the way to the end of the functions params before inserting the bisect code. Before ppxlib.0.36 this would have been `Pexp_fun (x, Pexp_newtype (t ... ))` but now it is `Pexp_function([arg x; newtype t]...)` so that might be causing this.

```diff
diff --git a/_build/default/test/instrument/control/function.t b/_build/default/test/instrument/control/function.t.corrected
index bfb978d..a39693a 100644
--- a/_build/default/test/instrument/control/function.t
+++ b/_build/default/test/instrument/control/function.t.corrected
@@ -59,6 +59,7 @@ Or-pattern.
   > EOF
   let _ =
    fun ___bisect_matched_value___ ->
+    ___bisect_visit___ 2;
     match ___bisect_matched_value___ with
     | None | Some _ ->
         (match[@ocaml.warning "-4-8-9-11-26-27-28-33"]
```

This looks similar, the representation before would have been `Pexp_fun(__bisect..., Pexp_function(cases)` whereas now it is `Pexp_function([__bisect...], Pfunction_cases(cases))`

aantron commented:
> ```diff
> ```diff
> -    let ___bisect_case_0___ x () =
> +    let ___bisect_case_0___ x =
> +     fun [@ppxlib.migration.stop_taking] () ->
>        ignore x;
>        ___bisect_post_visit___ 0 (print_endline "foo")
> -    and ___bisect_case_1___ y () =
> +    and ___bisect_case_1___ y =
> +     fun [@ppxlib.migration.stop_taking] () ->
>        ignore y;
>        ___bisect_post_visit___ 1 (print_endline "bar")
>      in
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> these failures are a known bug that is fixed upstream, see ocaml-ppx/ppxlib#604.
> ```

That looks like the wrong issue link but I was able to find https://github.com/ocaml-ppx/ppxlib/issues/597 from your hint, thank you. Leaving this comment here so that we have a link to that issue for future reading :)

patricoferris committed:
Catch function cases earlier in instrumentation

patricoferris commented:
Thanks @aantron for catching that! 

I have pushed a commit ebb3526 that fixes the control/function.t and control/newtype.t. The fix was to match more closely how the original instrumentation caught `let f = function A -> ...` sooner as there was a dedicated parsetree node for this. Now it is hidden within a `Pexp_function` with no parameters.

mbarbin commented:
Thank you all for the work on this PR!

In the meanwhile, I have found great value in using https://github.com/kit-ty-kate/opam-alpha-repository Thanks!

@kit-ty-kate may I request that you add a new preview package for bisect_ppx which would include the latest fix for https://github.com/aantron/bisect_ppx/issues/450 this would allow continue the 5.4 testing while things settles. I am not completely sure it's possible to find a candidate version though (just in case this is possible to find such a revision to build from). Thank you!

AllanBlanchard commented:
Apparently, the fix is not robust enough. It still does not work with the following example:
```ocaml
[@@@warning "-37"]
type 'a bt = (int * 'a) list
type x

type ('a, 'b) t =
  | B  : 'a bt -> ('a, 'b) t
  | C : ('a, 'b) t
  | D : ('a, 'b) t

let _action (f  : int -> 'a -> 'b -> 'b) (acc : 'b) : ('a, x) t -> 'b =
  function
  | C | D -> acc
  | B l -> List.fold_left (fun acc (cs,x) -> f cs x acc) acc l
[@@@warning "+37"]
```

Without coverage, it compiles, with bisect we get the following error:
```
File "file.ml", line 20, characters 13-16:
20 |   | C | D -> acc
                  ^^^
Error: The value acc has type 'b but an expression was expected of type
         ('a, x) t -> 'b
       The type variable 'b occurs inside ('a, x) t -> 'b
```

This is a reduced example taken from Frama-C. I can point out the actual file if it is useful.

NathanReb commented:
@AllanBlanchard thanks for the minimization, I confirm I can reproduce the error with it and am working on a fix!

NathanReb committed:
Add regression test for @AllanBlanchard instrumentation bug

Signed-off-by: Nathan Rebours <nathan.rebours@ocamlpro.com>

NathanReb committed:
Fix instrumentation of Pexp_function

Signed-off-by: Nathan Rebours <nathan.rebours@ocamlpro.com>

NathanReb commented:
Just opened https://github.com/patricoferris/bisect_ppx/pull/2 which should fix the error @AllanBlanchard !

kit-ty-kate commented:
@mbarbin done: https://github.com/kit-ty-kate/opam-alpha-repository/commit/c9bdcd12a7ceec641ed54122bcee0c3b65980231

mbarbin commented:
> @mbarbin done: [kit-ty-kate/opam-alpha-repository@c9bdcd1](https://github.com/kit-ty-kate/opam-alpha-repository/commit/c9bdcd12a7ceec641ed54122bcee0c3b65980231)

@kit-ty-kate Very nice. I re-run a few CIs to pick up this new pkg and I can confirm this fixed the bisect_ppx related issues with 5.4 I had seen in these PRs. Continuing with the 5.4 transition here. Thanks a lot, and thank you too to maintainers and contributors for the work!

NathanReb commented:
@aantron what are the remaining blockers on this PR?

patricoferris committed:
Additional fixes

patricoferris commented:
Thanks @NathanReb! Merged into this branch.

AllanBlanchard commented:
> Just opened [patricoferris#2](https://github.com/patricoferris/bisect_ppx/pull/2) which should fix the error @AllanBlanchard !

Tested today, we can compile and test Frama-C with this fix. There are a few micro-differences on some counters but it changes global coverage by less than 0.05%.

Thanks a lot for that!

aantron commented:
Thanks everyone! I will review it in its current state shortly, and if there are ~~no~~ any* obvious problems I'll try to fix them myself, then do a merge and release.

kit-ty-kate commented:
Gentile ping

aantron commented:
For transparency for what's going on on my end, right now I am handing over Dream to a team of maintainers (https://discuss.ocaml.org/t/dream-looking-for-maintainers-to-take-ownership/17607). I hope to complete the main part of that within a week (but subject to it being the holiday season for most people, though not me). After that I will either have enough attention resources to personally work on all the rest of my projects, including Bisect_ppx, since all of them together need much less attention than Dream, or I will start handing them over, too, probably starting with Bisect_ppx as the next-most significant one after Dream for the OCaml community. Thanks for still having patience, which I ended up testing considerably, though not intentionally!

patricoferris commented:
Thanks for the update @aantron, I think the https://github.com/ocaml-ppx organisation could be a home for bisect_ppx (if you don't already have one in mind e.g. https://github.com/ocaml-community). Let me know if there's anything I can do on my end :)

zshipko commented:
Is there any way I can help push this along? We are hoping to make a new irmin release but need ppxlib 0.36 support

davesnx commented:
Same, I would like to have ppxlib and bisect_ppx in https://github.com/melange-community/melange-testing-library/

patricoferris commented:
Hi @aantron,

I hope all is well. The offer above still stands, let me know if this is something you would like/how I can help.

Kakadu commented:
@aantron Gentle ping.

Kakadu commented:
> Hi @aantron,
> 
> I hope all is well. The offer above still stands, let me know if this is something you would like/how I can help.

This PR soon be 1 year old. Maybe we should prepare for forking?

Octachron commented:
@aantron , would you mind if I created `bisect_ppx_community` community version at https://github.com/ocaml-community to unlock the compatibility with recent versions of ppxlib and OCaml while you sort out what path you wish to take in term of maintenance for `bisect_ppx` ?

NathanReb commented:
If @aantron has a bit of time to check on this PR and answer here, they might as well add other maintainers (such as myself for instance) or transfer the project over to `ocaml-ppx`, I guess it'd be preferable to keep the main project over forking.

Kakadu commented:
I spend some time looking at this, and testsuite is not OK. Good new is that it compiles for 5.5 without a change.

1) There are many differences related to ocamlformat. The output differs because it breaks backward compatibility in every version. Maybe we should use not so fancy output from compiler itself, because it breaks pretty printing in more rare cases.
2) Some dune cram tests missing explicit dependency on the package `bisect-ppx`. Easy to fix. 
3) It looks like instrumentalization of for loops is broken. (See test/instrument/control/for.t). No idea when it started to work incorrectly.

kit-ty-kate commented:
I've opened https://github.com/ocaml-community/meta/issues/51. I propose waiting until OCaml 5.6 is about to be released to go ahead if we still haven't heard from the original author.
